stdlib/public/runtime/Win32Strings.cpp

```diff
@@ -0,0 +1,50 @@
+//===--- Win32Strings.cpp -------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Includes conversion functions for use with Windows API that produces or
+// consumes wide strings (wchar_t) which, on Windows, are UTF-16.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/Win32Strings.h"
+
+#if defined(_WIN32)
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <Windows.h>
+
+#include <cstdlib>
+
+using namespace swift;
+
+char *swift::swift_copyUTF8FromWide(wchar_t *str) {
+    int resultLength = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, str,
+                                           -1, nullptr, 0, nullptr, nullptr);
+    if (resultLength <= 0) {
+        return nullptr;
+    }
+
+    if (char *result = reinterpret_cast<char *>(malloc(resultLength))) {
+        resultLength = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, str,
+                                           -1, result, resultLength, nullptr,
+                                           nullptr);
+        if (resultLength <= 0) {
+            free(result);
+        } else {
+            return result;
+        }
+    }
+
+    return nullptr;
```

Comment:
Seems it would be simpler to write this as:

```
if (resultLength)
  return result;
free(result);
return nullptr;
```

Reply:
yoda.gif

Reply:
Do we know if the input `str` will always be null-terminated? Seems worth the defensive terminating null byte write on the happy path here.

Reply:
The Win32 API guarantees that on success, it has written the null character. The input is a C string, so… yes to that too?